### PR TITLE
[JUJU-430] Refactor DBAccessor Worker

### PR DESCRIPTION
**Pull request by SimonRichardson**
The worker is now refactored to allow for testing and building with cgo
and no-cgo implementations. This should drastically improve workflow. It
also ensures that the cgo changes do not leak outside of this folder for
now.

## QA steps (x86_64 only)

```sh
# If you don't have musl installed make sure to follow the hints emitted by the Makefile errors!
$ make go-build cgo-go-build
$ _build/linux_amd64/bin/juju bootstrap lxd test --no-gui --build-agent
$ juju enable-ha

# Wait for the machines to join the HA cluster...

# Connect to machine-0 and check the logs (CTRL+D to exit)
$ make juju-dqlite-repl
.open logs
SELECT entity, COUNT(*) from logs group by entity;

# Open a new terminal, connect to the other machines and ensure 
# that you can still access all logs by running the same commands as above
# 
# Use lxc list to find the  names of the other boxes
$ JUJU_CONTROLLER_BOX=juju-XXX-1 make juju-dqlite-repl
.open logs
SELECT entity, COUNT(*) from logs group by entity;

$ JUJU_CONTROLLER_BOX=juju-XXX-2 make juju-dqlite-repl
.open logs
SELECT entity, COUNT(*) from logs group by entity;

# Keep running the same queries on the different terminals and verify that the logs are in sync.
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -19,14 +19,14 @@ define MAIN_PACKAGES
   github.com/juju/juju/cmd/containeragent
   github.com/juju/juju/cmd/plugins/juju-metadata
   github.com/juju/juju/cmd/plugins/juju-wait-for
+  github.com/juju/juju/cmd/jujuc
```
Comment (SimonRichardson) [inline Makefile:22]: @jameinel this is now fixed: `11M Jan 12 12:02 jujuc` is back down to normal.

File: Makefile [hunk 2]
```diff
@@ -128,7 +128,7 @@ run-tests:
 	$(eval TMP := $(shell mktemp -d $${TMPDIR:-/tmp}/jj-XXX))
 	$(eval TEST_PACKAGES := $(shell go list $(PROJECT)/... | grep -v $(PROJECT)$$ | grep -v $(PROJECT)/vendor/ | grep -v $(PROJECT)/acceptancetests/ | grep -v $(PROJECT)/generate/ | grep -v mocks))
 	@echo 'go test -mod=$(JUJU_GOMOD_MODE) -tags "$(BUILD_TAGS) $(REQUIRED_BUILD_TAGS)" $(TEST_ARGS) $(CHECK_ARGS) -test.timeout=$(TEST_TIMEOUT) $$TEST_PACKAGES -check.v'
-	@TMPDIR=$(TMP) CGO_LDFLAGS_ALLOW=$(CGO_LDFLAGS_ALLOW) go test -mod=$(JUJU_GOMOD_MODE) -tags "$(BUILD_TAGS) $(REQUIRED_BUILD_TAGS)" $(TEST_ARGS) $(CHECK_ARGS) -test.timeout=$(TEST_TIMEOUT) $(TEST_PACKAGES) -check.v
+	@CGO_ENABLED=0 TMPDIR=$(TMP) go test -mod=$(JUJU_GOMOD_MODE) -tags "$(BUILD_TAGS) $(REQUIRED_BUILD_TAGS)" $(TEST_ARGS) $(CHECK_ARGS) -test.timeout=$(TEST_TIMEOUT) $(TEST_PACKAGES) -check.v
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply Makefile:131]: The offending line is here https://github.com/canonical/go-dqlite/blob/master/internal/bindings/build.go#L4, even if we use `// +build cgo` on that file it somehow still fails the build and is like `CGO_ENABLED=1` got automatically switched on. 

The CGO_ENABLED=0 only needs to be forced for `go test -v ./worker/dbaccessor` which is why it's on the make file. Running the following seems to yield the results we require.

```
➜ go test -v ./worker/dbaccessor
go build github.com/canonical/go-dqlite/internal/bindings: invalid flag in #cgo LDFLAGS: -Wl,-z,now
FAIL    github.com/juju/juju/worker/dbaccessor [build failed]
FAIL

➜ go test -v ./worker/pubsub
=== RUN   TestPackage

OK: 25 passed
--- PASS: TestPackage (0.19s)
PASS
ok      github.com/juju/juju/worker/pubsub      0.209s
```

File: worker/dbaccessor/worker_cgo.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	return a.dqliteApp.ID()
+}
+
+// Close the application node, releasing all resources it created.
+func (a *dqliteApp) Close() error {
+	return a.dqliteApp.Close()
+}
+
+// Repl returns a Repl worker from the underlying DB.
+func (a *dqliteApp) Repl(dbGetter DBGetter) (REPL, error) {
```
Comment (achilleasa) [inline worker/dbaccessor/worker_cgo.go:89]: Maybe `getREPL` to be consistent with the acronym?

File: worker/dbaccessor/worker_notcgo.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+// Open the dqlite database with the given name
+func (a *noopApp) Open(ctx context.Context, databaseName string) (*sql.DB, error) {
+	return nil, errors.NotSupportedf("db")
+}
+
+// Ready can be used to wait for a node to complete some initial tasks that are
+// initiated at startup.
+func (a *noopApp) Ready(ctx context.Context) error {
+	return nil
```
Comment (achilleasa) [inline worker/dbaccessor/worker_notcgo.go:30]: return nil, errors.NotSupportedf("db")

File: worker/dbaccessor/worker_notcgo.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// Handover transfers all responsibilities for this node (such has leadership
+// and voting rights) to another node, if one is available.
+//
+// This method should always be called before invoking Close(), in order to
+// gracefully shutdown a node.
+func (a *noopApp) Handover(ctx context.Context) error {
+	return nil
```
Comment (achilleasa) [inline worker/dbaccessor/worker_notcgo.go:39]: return nil, errors.NotSupportedf("db")

File: worker/dbaccessor/worker_notcgo.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// ID returns the dqlite ID of this application node.
+func (a *noopApp) ID() uint64 {
+	return 0
+}
+
+// Close the application node, releasing all resources it created.
+func (a *noopApp) Close() error {
+	return nil
```
Comment (achilleasa) [inline worker/dbaccessor/worker_notcgo.go:49]: return nil, errors.NotSupportedf("db")

File: worker/dbaccessor/worker_notcgo.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+package dbaccessor
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/juju/errors"
+)
+
+type noopApp struct{}
```
Comment (achilleasa) [inline worker/dbaccessor/worker_notcgo.go:15]: Should this perhaps be an unsupported or incompatible app instead? No-op is a bit ambiguous given that `Open` always errors.

**Review by achilleasa — APPROVED** [review]
LGTM. Some minor comments for naming conventions

File: worker/dbaccessor/worker_cgo.go [hunk 8]
```diff
@@ -85,8 +85,8 @@ func (a *dqliteApp) Close() error {
 	return a.dqliteApp.Close()
 }
 
-// Repl returns a Repl worker from the underlying DB.
-func (a *dqliteApp) Repl(dbGetter DBGetter) (REPL, error) {
+// GetRepl returns a Repl worker from the underlying DB.
+func (a *dqliteApp) GetRepl(dbGetter DBGetter) (REPL, error) {
```
Comment (achilleasa) [inline worker/dbaccessor/worker_cgo.go:89]: minor nit: `getREPL` as it is an acronym ;-)

Comment (SimonRichardson) [reply worker/dbaccessor/worker_cgo.go:89]: whoops!

**Pull request merged**